### ISSUE TITLE
Implement Theme Asset Qualifier

### ISSUE THREAD

**Issue by pdecostervgls** [unoplatform/uno#2197]
## What would you like to be added:
The theme qualifier for assets.

## Why is this needed:
So images can be different depending on the theme.

## For which Platform:

- [x] iOS
- [x] Android
- [x] WebAssembly
- [x] WebAssembly renderers for Xamarin.Forms
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

## Anything else we need to know?
As outlined in: https://docs.microsoft.com/en-us/windows/uwp/app-resources/images-tailored-for-scale-theme-contrast

**Comment by jeromelaban** [comment]
Thanks for the report!

**Comment by pdecostervgls** [comment]
@jeromelaban You're welcome, could you assign it to me?, since I am planning om picking this up as soon as possible, since I'm working on dark mode support for my app.

**Comment by pdecostervgls** [comment]
I'm having trouble on finding out where to start on this. I found https://github.com/unoplatform/uno/blob/f3d6bb85913bbf9ecf573861e04029531d9623fb/src/SourceGenerators/Uno.UI.Tasks/Assets/RetargetAssets.cs but I cannot get the logs to show up. @ghuntley could you help me get started on this?

**Comment by MartinZikmund** [comment]
@pdecostervgls The linked source code is just a build task, which makes sure the platform-specific assets get copied to the respective platforms during build. To achieve the `theme` qualifiers, there will be two components of the problem. 

First you need to find out how (and if even) the `theme-XYZ` qualifier is copied to the build output folder on Android/iOS/WASM in some form. I guess that the qualifier might be omitted from the file name on Android and iOS and copied as-is on WASM at the moment.  This will mean you will have to modify the `RetargetAssets` task to keep the qualifier in the file name or copy the assets to appropriate folders (in case of Android there is definitely a `Resources/drawable-???` qualifier for dark/light theme, for example).

The second part of the task will be to implement correct asset resolution on the target platforms. This might be quite easy on Android, as it could be handled automatically for you by the OS, but on iOS and WASM it might require additional work.

**Comment by pdecostervgls** [comment]
@MartinZikmund I have verified that assets with the theme qualifier do already get copied to the build output folder, so I guess only the asset resolution part is left. I could look at how the scale assets get resolved and work from there?

**Comment by MartinZikmund** [comment]
That sounds good! Let me know when you encounter something :-)

**Comment by pdecostervgls** [comment]
I cannot find it. I'm starting to suspect that the asset resolving is actually implemented in Xamarin instead.

**Comment by jeromelaban** [comment]
Assets resolving for iOS and Android is done by the platform, given the proper naming convention.

The assets retargeting is [done on Android by the AndroidResourcesConverter ](https://github.com/unoplatform/uno/blob/d9323226b6060eee1b7763b88b2a082da971510f/src/Uno.UWP/ApplicationModel/Resources/Core/AndroidResourceConverter.cs), called from the [RetargetAssets task](https://github.com/unoplatform/uno/blob/master/src/SourceGenerators/Uno.UI.Tasks/Assets/RetargetAssets.cs#L51-L62).

Adding dark theme support would probably be similar to the support for multiple DPIs.

**Comment by pdecostervgls** [comment]
I though that on iOS you had to create an assets catalog with the corresponding contents.json file to declare the DPIs. Other than retargeting the assets I cannot find anything creating a contents.json manifest for iOS.

**Comment by jeromelaban** [comment]
@pdecostervgls that's used for assets catalogs, not for bundle resources. But that may work differently for dark theme, there probably documentation on that.

**Comment by pdecostervgls** [comment]
@jeromelaban As far is I can find in the documentation, you cannot signify bundle resources for dark/light mode without wrapping them in a asset catalog.

**Comment by jeromelaban** [comment]
For iOS, that would make sense. Have you looked it up for Android as well ?

**Comment by pdecostervgls** [comment]
According to https://medium.com/androiddevelopers/appcompat-v23-2-daynight-d10f90c83e94 it supports the (renamed) asset qualifer, I can use the retargetassets for that too.

**Comment by pdecostervgls** [comment]
Come to think of it, I dont know for sure that the Image resolver will actually reference the xcassets entries that I plan on generating. Since the actual files do get retargetted as expected, wouldn't it be wiser to just override the Image Source resolver to try to match the image with the most fitting qualifiers at runtime? @jeromelaban @MartinZikmund

**Comment by jeromelaban** [comment]
@pdecostervgls for iOS, it's possible to generate the proper files and add them as ItemGroups so that the iOS tooling picks those up.

For the other platforms, and Wasm specifically, we'll have to handle that at runtime in all cases.

**Comment by pdecostervgls** [comment]
@MartinZikmund I unassigned myself because I lack the experience with VS ResourceGeneration to complete this. Would you like to pick it up? I got some work done (UI tests and set-up) at [this branch](http://www.github.com/pdecostervgls/Uno/tree/dev/pdeco/theme-qualifier)

**Comment by pdecostervgls** [comment]
Is this closed because it is implemented? If so, i'd like to see the code, for education purposes :)

**Comment by jeromelaban** [comment]
It's not, it was closed incorrectly. Sorry about that :)

**Comment by hawkerm** [comment]
Ah, I just hit this with WASM trying to use an asset like `Square150x150Logo.theme-light_scale-100.png` and referencing in an image directly:

```xml
<Image Width="150"
             Height="150"
             Source="ms-appx:///Assets/Square150x150Logo.png" />
```

Saw there's a workaround that Labs did using ThemeDictionaries, see here: https://github.com/CommunityToolkit/Labs-Windows/blob/3669d4c52eafb2bcb9b1d782d85c1a49111e175f/common/CommunityToolkit.Labs.Shared/Pages/GettingStartedPage.xaml

**Comment by jeromelaban** [comment]
That's correct. At this time, [Wasm only supports assets scaling](https://github.com/unoplatform/uno/blob/e749670ebd8541b120f8a160e66c4762e0e8b88a/src/Uno.UI/UI/Xaml/Media/Imaging/BitmapImage.wasm.cs#L103), and a theme resources workaround is the best way.

**Comment by nickrandolph** [comment]
@jeromelaban @francoistanguay from a compatibility point of view, it would be good to get this onto the backlog. Having theme qualifiers would save the need for workarounds to handle theme based images.

**Comment by xperiandri** [comment]
Is there any work going on this?

**Comment by MartinZikmund** [comment]
# Repro generator

This issue was classified as not reproducible with a code sample or test.

**Reason:** This is a feature request for theme-qualified asset loading (different images for light/dark themes via resource qualifiers). There is no specific broken behavior to demonstrate — it is a new feature for the asset resolution pipeline.
